File: ch1/isunique.cc
// is_unique is an algorithm which determines if a string has all
// unique characters
// is_unique2 does the same as is_unique but does not use additional
// data structures

#include <iostream>
#include <string>
#include <map>

bool is_unique(std::string str) {
  int len = str.length();
  std::map<char, bool> chars;

  for (int i = 0; i < len; i++) {
    if (chars.count(str[i]) != 0) return false;

    else {
      chars[str[i]] = true;
    }
  }

  return true;
}

bool is_unique2(std::string str) {
  int len = str.length();

  for (int i = 0; i < len; i++) {
    for (int j = i; j < len; j++) {
      if (str[i] == str[j]) return false;
    }
  }

  return true;
}

int main() {
  std::string str1 = "abcde";
  std::string str2 = "abcdee";

  std::cout << is_unique(str1) << is_unique(str2) << std::endl;
}

File: ch1/is_unique.py
def is_unique(s):
    charset = set({})

    for char in s:
        if char in charset:
            return False
        
        charset.add(char)

    return True

def main():
    str1 = "abcde"
    str2 = "babcde"

    print(is_unique(str1))
    print(is_unique(str2))

if __name__ == "__main__":
    main()